### ISSUE TITLE
Simplify actor API to simplify and improve efficiency

### ISSUE THREAD

**Issue by lulf** [drogue-iot/drogue-device#20]
I'm going to focus on the NotifyHandler trait here, but arguments apply equally to the RequestHandler trait.

At present, an Actor have the ability to handle multiple message types by implementing a NotifyHandler for a given message type. This provides a lot of flexibility when implementing Actors, but there are a few implications of this model:

* Actors can only process a single message at a time, so a notification handler must not block. When implementing actors, care must be taken to ensure processing of one message does not potentially block another, which can be
counter-intuitive in a complex driver having multiple NotifyHandler implementations spread throughout the source code: any of these handlers cannot use an "unbounded" await if it also expects some event to happen that "unblocks" the awaited request.

* ActorContext must Box incoming requests/messages, because it can be any type. This causes a need for heap to be used to store the messages in the actor context, rather than being able to allocate space for potential requests at compile time.

* Most driver implementation packages have a separate component for handling interrupts (so that an interrupt cannot be blocked), an actor component that handles the incoming requests that may block, and sometimes a controller component in the middle that can do the 'background processing' to prevent the 'actor' component from blocking the processing logic.

With the above limitations, I think the current APIs and traits provide "too much" flexibility that cannot be utilized anyway.

### An alternative model

Use associated types on actors to specify a RequestType and a ResponseType for an actor (default `()`). The Actor trait could specify a default "on_request" implementation for this type  actors that don't handle any incoming messages. Actors that handle messages can specify both the types and the handler implementation for those types. 

Notify-style messages are also of the RequestType, but the Address<> API is still not async. The ActorContext will just disregard the response value.

Advantages:

* Actor implementation can more easily be reasoned about since it is clear that only one on_request() can be blocking at a time. 
* No need to think about notify vs request from an implementation: A notify is just a request where response is ignored.
* Should make way for some code and memory usage savings, I think ActorContext can now reserve space at compile time for incoming messages for a given actor.

The API towards the users is the Address<>, which encapsulates the actual message types sent to the actor, so all APIs in drogue-device would stay the same, and in the event of handling multiple requests, an enum would be used to support different message types.

Disadvantages:
* Complex actors with request types and responses of different sizes will be forced to store the largest sizes. However, given that we have a way to deal with references to data now, I think it should be fine.

@bobmcwhirter ^^

**Comment by bobmcwhirter** [comment]
+1!

**Comment by lulf** [comment]
@bobmcwhirter 

After doing several rewrites now and some thinking, the following issues is what I've run into:

- Requests that hold references need a lifetime specifier. However, as generic associated types is not supported in Rust (`type Request<'a> = UartRequest<'a>`), the lifetime specifier either need to be 'static (or lifetime of Actor). 

- Extending behavior of existing drivers to hook into platform-specific code like GpioteEvent -> Button might not be possible because Button is hardcoded to a single request type. In general, this means that an Actor's request type must cover all possible future messages that can be handled by that Actor, which limits extensibility.

- EventBus can no longer be an Actor, because it must support different message types.

I still think the NotifyHandler -> RequestHandler makes sense though, so I vote for keeping that. I need to rethink a bit if
there is a middle ground here that would still avoid heap alloc, but I fear that it entails the need to know the exact size of all types, so it inherently limits extensibility.

I will leave this and rethink over the weekend.

**Comment by lulf** [comment]
@bobmcwhirter @jcrossley3 

I got the first part of this to work. Essentially each actor has associated types Request and Response. This requires most of the existing drivers to use an enum in cases where they support multiple types. The APIs are also simplified a bit, since an API is essentially and Actor with specific Request and Response types. I also re-introduced the on_notify as a method on Actor. Both on_request and on_notify have default implementations (requires Response: Default), so they can be optional. The end result is something like this:

```
impl Actor for MyActor {
    type Request = MyRequest;
    type Response = ();

    fn on_request(self, request: Self::Request) -> Response<Self> {
        match request {
            MyRequest::Write => {
            }
            MyRequest::Read => {
            }
        }
        Response::immediate(self, ())
    }

    fn on_notify(self, request: Self::Request) -> Completion<Self> {
        match request {
            MyRequest::Write => {
            }
            MyRequest::Read=> {
            }
        }
        Response::immediate(self, ())
    }
}
```

In the event that the request holds references to data, the Address<MyActor> impl needs to do a transmute: this is necessary because associated types cannot specify lifetime and generic paramters to its types (i.e. Request<'a> = MyRequest<'a>) yet.

A weakness is that one cannot extend a driver with new message handlers, which was previously possible. I.e. the nRF-specific GPIO implementation contains a `NotifyHandler<GpioteEvent> for Button`, which allowed the button driver to receive interrupts 'via' the nRF GPIO interrupt.

Another quirk is the EventBus, which can no longer be an Actor, because it needs to handle multiple message types. The implication of this is that the EventBus need to dispatch events to the device directly, and cannot dispatch them to a generic "NotifyHandler" for instance.

Moreover, the code size was about 10% smaller with this improvement, not as much as I'd hoped.

The second part relates to no-allocifying the `Response` and `Completion` types that currently use Box/alloc to store deferred futures. My initial attempt ended up adding addition associated types to the Actor trait to define the futures that could be returned from an Actor. However, this complicates actor implementations greatly: instead of using `Response::defer(async move..)`, the driver needs to implement a "DeferredFuture" that must handle all the cases of manual polling. It also requires a whole lot of rework of the 'response' path.

In the end, this gets gets rid of alloc completely within system, and the additional size reduction from this is another 30% on top of the initial 10% for the first part.

I think the first part could've been worth it if one would be able to complete the second part, making the entire thing no-alloc. However, all these quirks makes me reconsider my original proposal, because its a lot more limiting than I first anticipated.

I propose that we do the following:

* Revert my previous NotificationHandler -> RequestHandler change.
* Store current changes in a github branch referenced from this issue.
* Revisit this topic using an alternate approach. One idea I had in mind was to use attribute macros on message types that would magically compose an enum containing all possible message types, and an Into/From<GlobalEnum> for them so that we could still get no-alloc for the stored messages.
